File: app/src/main/java/hk/collaction/freehkkai/ui/fragment/SettingsFragment.java
package hk.collaction.freehkkai.ui.fragment;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.v7.preference.Preference;
import android.view.View;

import com.afollestad.materialdialogs.MaterialDialog;
import com.blankj.utilcode.util.AppUtils;

import hk.collaction.freehkkai.BuildConfig;
import hk.collaction.freehkkai.C;
import hk.collaction.freehkkai.R;


public class SettingsFragment extends BasePreferenceFragment {

	private SharedPreferences settings;
	private Preference prefFontVersion;

	public static SettingsFragment newInstance() {
		return new SettingsFragment();
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		addPreferencesFromResource(R.xml.pref_general);
	}

	@Override
	public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);

		settings = PreferenceManager.getDefaultSharedPreferences(mContext);

		/* Set version */
		Preference prefVersion = findPreference("pref_version");
		prefVersion.setSummary(AppUtils.getAppVersionName());

		findPreference("pref_report").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
			@Override
			public boolean onPreferenceClick(Preference preference) {
				Intent intent = new Intent(Intent.ACTION_SEND);

				String text = "Android Version: " + Build.VERSION.RELEASE + "\n";
				text += "SDK Level: " + String.valueOf(Build.VERSION.SDK_INT) + "\n";
				text += "Version: " + AppUtils.getAppVersionName() + "\n";
				text += "Brand: " + Build.BRAND + "\n";
				text += "Model: " + Build.MODEL + "\n\n\n";

				intent.setType("message/rfc822");
				intent.putExtra(Intent.EXTRA_EMAIL, new String[]{BuildConfig.CONTACT_EMAIL});
				intent.putExtra(Intent.EXTRA_SUBJECT, "自由香港楷書回報問題");
				intent.putExtra(Intent.EXTRA_TEXT, text);
				startActivity(intent);
				return false;
			}
		});

		findPreference("pref_rate").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
			@Override
			public boolean onPreferenceClick(Preference preference) {
				Uri uri = Uri.parse("market://details?id=" + mContext.getPackageName());
				Intent intent = new Intent(Intent.ACTION_VIEW, uri);
				startActivity(intent);
				return false;
			}
		});

		findPreference("pref_share").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
			@Override
			public boolean onPreferenceClick(Preference preference) {
				Intent intent = new Intent();
				intent.setAction(Intent.ACTION_SEND);
				intent.putExtra(Intent.EXTRA_TEXT, "下載「自由香港楷書」程式，就可以查詢支援超過 4700 個香港教育局楷書參考寫法，解決因為「電腦輸入法」而令學生 / 家長 / 教師混淆而寫錯字的問題。\n\n" + "https://play.google.com/store/apps/details?id=" + mContext.getPackageName());
				intent.setType("text/plain");
				startActivity(Intent.createChooser(intent, "分享此程式"));
				return false;
			}
		});

		findPreference("pref_author").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
			@Override
			public boolean onPreferenceClick(Preference preference) {
				Uri uri = Uri.parse("market://search?q=pub:\"Collaction 小隊\"");
				Intent intent = new Intent(Intent.ACTION_VIEW, uri);
				startActivity(intent);
				return false;
			}
		});

		findPreference("pref_collaction").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
			@Override
			public boolean onPreferenceClick(Preference preference) {
				Uri uri = Uri.parse("https://www.collaction.hk/s/collactionopensource");
				Intent intent = new Intent(Intent.ACTION_VIEW, uri);
				startActivity(intent);
				return false;
			}
		});

		findPreference("pref_hkfreekai").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
			@Override
			public boolean onPreferenceClick(Preference preference) {
				Uri uri = Uri.parse("https://www.collaction.hk/s/freehkfonts");
				Intent intent = new Intent(Intent.ACTION_VIEW, uri);
				startActivity(intent);
				return false;
			}
		});

		prefFontVersion = findPreference("pref_font_version");
		prefFontVersion.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
			@Override
			public boolean onPreferenceClick(Preference preference) {
				String fontVersion = settings.getString("pref_font_version", "fonts/freehkkai_4700.ttf");
				int a = 0;
				switch (fontVersion) {
					case "fonts/freehkkai_extended.ttf":
						a = 1;
						break;
				}

				MaterialDialog.Builder dialog = new MaterialDialog.Builder(mContext)
						.title("切換字型檔案版本")
						.items(R.array.font_version_array)
						.itemsCallbackSingleChoice(a, new MaterialDialog.ListCallbackSingleChoice() {
							@Override
							public boolean onSelection(MaterialDialog dialog, View itemView, int which, CharSequence text) {
								switch (which) {
									case 0:
										settings.edit().putString(C.PREF_FONT_VERSION, "fonts/freehkkai_4700.ttf").apply();
										break;
									case 1:
										settings.edit().putString(C.PREF_FONT_VERSION, "fonts/freehkkai_extended.ttf").apply();
										break;
								}
								setFontVersionSummary();
								return false;
							}
						})
						.negativeText(R.string.ui_cancel);
				dialog.show();

				mContext.setResult(Activity.RESULT_OK);

				return false;
			}
		});
		setFontVersionSummary();
	}

	void setFontVersionSummary() {
		prefFontVersion.setSummary(C.getCurrentFontName(mContext));
	}
}

File: app/src/main/java/hk/collaction/freehkkai/ui/settings/SettingsFragment.kt
package hk.collaction.freehkkai.ui.settings

import android.app.Activity
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Build
import android.os.Bundle
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.list.listItemsSingleChoice
import com.blankj.utilcode.util.AppUtils
import hk.collaction.freehkkai.BuildConfig
import hk.collaction.freehkkai.R
import hk.collaction.freehkkai.util.Utils.PREF_FONT_VERSION
import hk.collaction.freehkkai.util.Utils.getCurrentFontName

class SettingsFragment : PreferenceFragmentCompat() {
    private lateinit var sharedPreferences: SharedPreferences
    private var prefFontVersion: Preference? = null

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.pref_general)

        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
        /* Set version */
        val prefVersion = findPreference<Preference>("pref_version")
        prefVersion!!.summary = AppUtils.getAppVersionName()
        findPreference<Preference>("pref_report")?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val intent = Intent(Intent.ACTION_SEND)
            var text = "Android Version: " + Build.VERSION.RELEASE + "\n"
            text += "SDK Level: " + Build.VERSION.SDK_INT.toString() + "\n"
            text += "Version: " + AppUtils.getAppVersionName() + "\n"
            text += "Brand: " + Build.BRAND + "\n"
            text += "Model: " + Build.MODEL + "\n\n\n"
            intent.type = "message/rfc822"
            intent.putExtra(Intent.EXTRA_EMAIL, arrayOf(BuildConfig.CONTACT_EMAIL))
            intent.putExtra(Intent.EXTRA_SUBJECT, "自由香港楷書回報問題")
            intent.putExtra(Intent.EXTRA_TEXT, text)
            startActivity(intent)
            false
        }
        findPreference<Preference>("pref_rate")?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val uri = Uri.parse("market://details?id=" + context?.packageName)
            val intent = Intent(Intent.ACTION_VIEW, uri)
            startActivity(intent)
            false
        }
        findPreference<Preference>("pref_share")?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val intent = Intent()
            intent.action = Intent.ACTION_SEND
            intent.putExtra(Intent.EXTRA_TEXT, "下載「自由香港楷書」程式，就可以查詢支援超過 4700 個香港教育局楷書參考寫法，解決因為「電腦輸入法」而令學生 / 家長 / 教師混淆而寫錯字的問題。\n\n" + "https://play.google.com/store/apps/details?id=" + context?.packageName)
            intent.type = "text/plain"
            startActivity(Intent.createChooser(intent, "分享此程式"))
            false
        }
        findPreference<Preference>("pref_author")?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val uri = Uri.parse("market://search?q=pub:\"Collaction 小隊\"")
            val intent = Intent(Intent.ACTION_VIEW, uri)
            startActivity(intent)
            false
        }
        findPreference<Preference>("pref_collaction")?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val uri = Uri.parse("https://www.collaction.hk/s/collactionopensource")
            val intent = Intent(Intent.ACTION_VIEW, uri)
            startActivity(intent)
            false
        }
        findPreference<Preference>("pref_hkfreekai")?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val uri = Uri.parse("https://www.collaction.hk/s/freehkfonts")
            val intent = Intent(Intent.ACTION_VIEW, uri)
            startActivity(intent)
            false
        }
        findPreference<Preference>("pref_privacy")?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val uri = Uri.parse("https://www.collaction.hk/about/privacy")
            val intent = Intent(Intent.ACTION_VIEW, uri)
            startActivity(intent)
            false
        }
        prefFontVersion = findPreference("pref_font_version")
        prefFontVersion?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val a = when (sharedPreferences.getString(PREF_FONT_VERSION, "4700")) {
                "7000" -> 1
                else -> 0
            }
            activity?.let { activity ->
                MaterialDialog(activity)
                        .title(text = "切換字型檔案版本")
                        .listItemsSingleChoice(
                                R.array.font_version_array,
                                initialSelection = a,
                                waitForPositiveButton = false
                        ) { dialog, index, _ ->
                            when (index) {
                                0 -> sharedPreferences.edit().putString(PREF_FONT_VERSION, "4700").apply()
                                1 -> sharedPreferences.edit().putString(PREF_FONT_VERSION, "7000").apply()
                            }
                            setFontVersionSummary()
                            dialog.dismiss()
                        }
                        .negativeButton(R.string.ui_cancel)
                        .show()

                activity.setResult(Activity.RESULT_OK)
            }
            false
        }
        setFontVersionSummary()
    }

    private fun setFontVersionSummary() {
        context?.let { context ->
            prefFontVersion?.setSummary(getCurrentFontName(context))
        }
    }

    companion object {
        fun newInstance(): SettingsFragment {
            return SettingsFragment()
        }
    }
}